File: test_stage2_qa.py
#!/usr/bin/env python3
"""
Stage 2 QA Test Script - Executions + Trading Accounts
Tests CRUD operations for executions and trading_accounts entities
"""

import requests
import json
import time
from datetime import datetime

BASE_URL = "http://localhost:8080"

def login():
    """Login as admin and get session"""
    session = requests.Session()
    login_data = {
        "username": "admin",
        "password": "admin123"
    }
    response = session.post(f"{BASE_URL}/api/auth/login", json=login_data)
    if response.status_code != 200:
        raise Exception(f"Login failed: {response.status_code}")

    # Store token in session headers for subsequent requests
    response_data = response.json()
    if 'data' in response_data and 'access_token' in response_data['data']:
        session.headers.update({'Authorization': f'Bearer {response_data["data"]["access_token"]}'})
    elif 'access_token' in response_data:
        session.headers.update({'Authorization': f'Bearer {response_data["access_token"]}'})

    return session

def test_entity_crud(session, entity_name, test_data):
    """Test CRUD operations for an entity"""
    print(f"\n🧪 Testing {entity_name} CRUD operations:")

    # CREATE
    print("  📝 CREATE operation...")
    create_response = session.post(f"{BASE_URL}/api/{entity_name}/", json=test_data)
    print(f"    Status: {create_response.status_code}")
    if create_response.status_code not in [200, 201]:
        print(f"    ❌ CREATE failed: {create_response.text}")
        return False

    created_data = create_response.json()
    record_id = created_data.get('data', {}).get('id') or created_data.get('id')
    if not record_id:
        print("    ❌ No ID returned from CREATE")
        return False
    print(f"    ✅ Created record ID: {record_id}")

    # READ
    print("  📖 READ operation...")
    read_response = session.get(f"{BASE_URL}/api/{entity_name}/{record_id}")
    print(f"    Status: {read_response.status_code}")
    if read_response.status_code != 200:
        print(f"    ❌ READ failed: {read_response.text}")
        return False
    print("    ✅ READ successful")

    # UPDATE
    print("  ✏️  UPDATE operation...")
    update_data = test_data.copy()
    update_data['id'] = record_id
    if 'notes' in update_data:
        update_data['notes'] = f"Updated notes - {datetime.now().isoformat()}"

    update_response = session.put(f"{BASE_URL}/api/{entity_name}/{record_id}", json=update_data)
    print(f"    Status: {update_response.status_code}")
    if update_response.status_code not in [200, 204]:
        print(f"    ❌ UPDATE failed: {update_response.text}")
        return False
    print("    ✅ UPDATE successful")

    # DELETE
    print("  🗑️  DELETE operation...")
    delete_response = session.delete(f"{BASE_URL}/api/{entity_name}/{record_id}")
    print(f"    Status: {delete_response.status_code}")
    if delete_response.status_code not in [200, 204]:
        print(f"    ❌ DELETE failed: {delete_response.text}")
        return False
    print("    ✅ DELETE successful")

    return True

def main():
    print("🚀 Stage 2 QA Test - Executions + Trading Accounts")
    print(f"Base URL: {BASE_URL}")
    print(f"Timestamp: {datetime.now().isoformat()}")

    try:
        # Login
        session = login()
        print("✅ Login successful")

        results = {}

        # Test Trading Accounts
        trading_account_data = {
            "name": f"Test Account QA {int(time.time())}",
            "currency_id": 1,  # USD
            "status": "open",
            "opening_balance": 10000.0,
            "notes": "QA Test Record - Safe to delete"
        }

        results['trading_accounts'] = test_entity_crud(session, 'trading-accounts', trading_account_data)

        # Test Executions - need valid ticker and trading account
        # First get/create required dependencies
        tickers_response = session.get(f"{BASE_URL}/api/tickers/")
        if tickers_response.status_code == 200:
            tickers = tickers_response.json().get('data', [])
            if tickers:
                ticker_id = tickers[0]['id']
                print(f"  📊 Using existing ticker ID: {ticker_id}")

                # Get trading accounts
                accounts_response = session.get(f"{BASE_URL}/api/trading-accounts/")
                if accounts_response.status_code == 200:
                    accounts = accounts_response.json().get('data', [])
                    if accounts:
                        account_id = accounts[0]['id']
                        print(f"  📊 Using existing trading account ID: {account_id}")

                        execution_data = {
                            "ticker_id": ticker_id,
                            "trading_account_id": account_id,
                            "action": "buy",
                            "quantity": 100,
                            "price": 50.0,
                            "date": datetime.now().isoformat(),
                            "fee": 5.0,
                            "source": "manual",
                            "notes": "QA Test Record - Safe to delete"
                        }

                        results['executions'] = test_entity_crud(session, 'executions', execution_data)
                    else:
                        print("❌ No trading accounts found for executions test")
                        results['executions'] = False
                else:
                    print(f"❌ Failed to get trading accounts: {accounts_response.status_code}")
                    results['executions'] = False
            else:
                print("❌ No tickers found for executions test")
                results['executions'] = False
        else:
            print(f"❌ Failed to get tickers: {tickers_response.status_code}")
            results['executions'] = False

        # Summary
        print("\n📊 Test Results Summary:")
        print(f"  Trading Accounts: {'✅ PASS' if results.get('trading_accounts') else '❌ FAIL'}")
        print(f"  Executions: {'✅ PASS' if results.get('executions') else '❌ FAIL'}")

        total_passed = sum(1 for result in results.values() if result)
        total_tests = len(results)

        print(f"\n🎯 Overall: {total_passed}/{total_tests} entities passed CRUD tests")

        if total_passed == total_tests:
            print("🎉 Stage 2 QA: ALL TESTS PASSED! ✅")
            return True
        else:
            print("⚠️  Stage 2 QA: Some tests failed")
            return False

    except Exception as e:
        print(f"❌ Test execution failed: {e}")
        return False

if __name__ == "__main__":
    success = main()
    exit(0 if success else 1)

File: test_stage2_qa.js
/**
 * Stage 2 QA Test - Executions + Trading Accounts
 * Focused QA run with Logger + Network evidence capture
 */

const fs = require('fs');
const path = require('path');

// Clear previous debug log
const logPath = path.join(__dirname, '.cursor', 'debug.log');
if (fs.existsSync(logPath)) {
    fs.unlinkSync(logPath);
}

// Simulate window and Logger
global.window = {
    Logger: {
        info: function(msg, data) {
            const entry = {
                id: `log_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
                timestamp: Date.now(),
                location: 'test_stage2_qa.js:Logger',
                message: msg,
                data: data || {},
                sessionId: 'stage2_qa_test',
                runId: 'stage2_qa_run_001',
                hypothesisId: 'qa_validation'
            };
            const logLine = JSON.stringify(entry) + '\n';
            fs.appendFileSync(logPath, logLine);
            console.log('ℹ️', msg, data ? JSON.stringify(data).substring(0, 100) + '...' : '');
        },
        error: function(msg, data) {
            const entry = {
                id: `log_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
                timestamp: Date.now(),
                location: 'test_stage2_qa.js:Logger.error',
                message: msg,
                data: data || {},
                sessionId: 'stage2_qa_test',
                runId: 'stage2_qa_run_001',
                hypothesisId: 'qa_validation'
            };
            const logLine = JSON.stringify(entry) + '\n';
            fs.appendFileSync(logPath, logLine);
            console.log('❌', msg, data ? JSON.stringify(data).substring(0, 100) + '...' : '');
        },
        warn: function(msg, data) {
            const entry = {
                id: `log_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
                timestamp: Date.now(),
                location: 'test_stage2_qa.js:Logger.warn',
                message: msg,
                data: data || {},
                sessionId: 'stage2_qa_test',
                runId: 'stage2_qa_run_001',
                hypothesisId: 'qa_validation'
            };
            const logLine = JSON.stringify(entry) + '\n';
            fs.appendFileSync(logPath, logLine);
            console.log('⚠️', msg, data ? JSON.stringify(data).substring(0, 100) + '...' : '');
        }
    }
};

// Mock fetch to simulate API calls and capture network
let capturedRequests = [];
global.fetch = async (url, options = {}) => {
    const method = options.method || 'GET';
    const body = options.body ? JSON.parse(options.body) : null;

    capturedRequests.push({
        url,
        method,
        headers: options.headers || {},
        body,
        timestamp: new Date().toISOString()
    });

    window.Logger.info(`Network: ${method} ${url}`, { body });

    // Mock responses
    if (url.includes('/api/trading-accounts') && method === 'POST') {
        return {
            ok: true,
            status: 201,
            json: async () => ({ id: 100, ...body })
        };
    }

    if (url.includes('/api/trading-accounts') && method === 'PUT') {
        return {
            ok: true,
            status: 200,
            json: async () => ({ id: parseInt(url.split('/').pop()), ...body })
        };
    }

    if (url.includes('/api/executions') && method === 'POST') {
        return {
            ok: true,
            status: 201,
            json: async () => ({ id: 200, ...body })
        };
    }

    if (url.includes('/api/executions') && method === 'PUT') {
        return {
            ok: true,
            status: 200,
            json: async () => ({ id: parseInt(url.split('/').pop()), ...body })
        };
    }

    if (url.includes('/api/trading-accounts') && method === 'GET') {
        return {
            ok: true,
            json: async () => [{ id: 1, name: 'Test Account', status: 'open' }]
        };
    }

    if (url.includes('/api/tickers') && method === 'GET') {
        return {
            ok: true,
            json: async () => [{ id: 1, symbol: 'AAPL', name: 'Apple Inc.' }]
        };
    }

    if (url.includes('/api/currencies') && method === 'GET') {
        return {
            ok: true,
            json: async () => [{ id: 1, code: 'USD', name: 'US Dollar' }]
        };
    }

    return {
        ok: false,
        status: 404,
        json: async () => ({ error: 'Not found' })
    };
};

// Import field maps
const fieldMaps = {
    trading_account: {
        required: ['name', 'currency_id'],
        fields: {
            name: { id: '#accountName', type: 'text', required: true },
            currency_id: { id: '#accountCurrency', type: 'int', required: true },
            opening_balance: { id: '#accountOpeningBalance', type: 'number', default: 0 },
            status: { id: '#accountStatus', type: 'text', default: 'open' },
            notes: { id: '#accountNotes', type: 'rich-text', default: null }
        },
        modalId: 'tradingAccountsModal'
    },
    execution: {
        required: ['ticker_id', 'trading_account_id', 'action', 'quantity', 'price', 'date'],
        fields: {
            ticker_id: { id: '#executionTicker', type: 'int', required: true },
            trading_account_id: { id: '#executionAccount', type: 'int', required: true },
            action: { id: '#executionType', type: 'text', required: true, default: 'buy' },
            quantity: { id: '#executionQuantity', type: 'number', required: true, default: 100 },
            price: { id: '#executionPrice', type: 'number', required: true, default: 100 },
            date: { id: '#executionDate', type: 'datetime-local', required: true },
            fee: { id: '#executionCommission', type: 'number', default: 0 },
            source: { id: '#executionSource', type: 'text', default: 'manual' },
            external_id: { id: '#executionExternalId', type: 'text', default: null },
            notes: { id: '#executionNotes', type: 'rich-text', default: null },
            realized_pl: { id: '#executionRealizedPL', type: 'number', default: null },
            mtm_pl: { id: '#executionMTMPL', type: 'number', default: null },
            trade_id: { id: '#executionTradeId', type: 'int', default: null }
        },
        modalId: 'executionsModal'
    }
};

// Mock UnifiedPayloadBuilder
const UnifiedPayloadBuilder = {
    validTickers: ['PLTR', 'AAPL', 'TSLA', 'MSFT', 'QQQ'],
    activeTradingAccountId: 1,

    build: function(entityType, fieldMap, isUpdate = false) {
        return this.generateTestData(entityType, fieldMap, isUpdate);
    },

    generateTestData: function(entityType, fieldMap, isUpdate = false) {
        const testData = {};

        for (const [fieldName, fieldConfig] of Object.entries(fieldMap.fields)) {
            testData[fieldName] = this.generateFieldValue(fieldName, fieldConfig, entityType, isUpdate);
        }

        this.applyEntitySpecificOverrides(testData, entityType, isUpdate);

        return testData;
    },

    generateFieldValue: function(fieldName, fieldConfig, entityType, isUpdate) {
        if (['trading_account_id', 'ticker_id', 'currency_id'].includes(fieldName)) {
            return 1; // Mock resolved ID
        }

        switch (fieldConfig.type) {
            case 'text':
                return fieldConfig.default || `Test ${fieldName}`;
            case 'number':
                return fieldConfig.default || 100;
            case 'int':
                return fieldConfig.default || 1;
            case 'datetime-local':
                return new Date().toISOString().slice(0, 16);
            case 'date':
                return new Date().toISOString().split('T')[0];
            case 'rich-text':
                return fieldConfig.default || 'Test notes';
            case 'bool':
                return fieldConfig.default || false;
            default:
                return fieldConfig.default || `default_${fieldName}`;
        }
    },

    applyEntitySpecificOverrides: function(testData, entityType, isUpdate) {
        switch (entityType) {
            case 'execution':
                testData.trading_account_id = 1;
                testData.ticker_id = 1;
                break;
            case 'trading_account':
                testData.currency_id = 1;
                break;
        }
    }
};

// Mock UnifiedCRUDService
const UnifiedCRUDService = {
    async create(entityType, data) {
        window.Logger.info(`CRUD Create: ${entityType}`, { data });

        const endpoint = entityType === 'trading_account' ? '/api/trading-accounts/' : '/api/executions/';

        const response = await fetch(endpoint, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(data)
        });

        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.message || 'Create failed');
        }

        return await response.json();
    },

    async update(entityType, id, data) {
        window.Logger.info(`CRUD Update: ${entityType} ${id}`, { data });

        const endpoint = entityType === 'trading_account' ? `/api/trading-accounts/${id}` : `/api/executions/${id}`;

        const response = await fetch(endpoint, {
            method: 'PUT',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(data)
        });

        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.message || 'Update failed');
        }

        return await response.json();
    },

    async delete(entityType, id) {
        window.Logger.info(`CRUD Delete: ${entityType} ${id}`);

        const endpoint = entityType === 'trading_account' ? `/api/trading-accounts/${id}` : `/api/executions/${id}`;

        const response = await fetch(endpoint, {
            method: 'DELETE'
        });

        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.message || 'Delete failed');
        }

        return { success: true };
    }
};

// Test function that mimics the dashboard behavior
async function runStage2QA() {
    console.log('🚀 Starting Stage 2 QA Test - Executions + Trading Accounts\n');

    const entities = ['trading_account', 'execution'];
    let passed = 0;
    let failed = 0;

    for (const entity of entities) {
        console.log(`📋 Testing ${entity}...`);

        try {
            const fieldMap = fieldMaps[entity];

            // Test Create
            console.log(`  🔄 Testing ${entity} CREATE...`);
            const createData = UnifiedPayloadBuilder.build(entity, fieldMap, false);
            const createResult = await UnifiedCRUDService.create(entity, createData);

            if (createResult.id) {
                console.log(`  ✅ ${entity} CREATE: PASSED (ID: ${createResult.id})`);
                passed++;

                // Test Update
                console.log(`  🔄 Testing ${entity} UPDATE...`);
                const updateData = UnifiedPayloadBuilder.build(entity, fieldMap, true);
                const updateResult = await UnifiedCRUDService.update(entity, createResult.id, updateData);

                if (updateResult.id) {
                    console.log(`  ✅ ${entity} UPDATE: PASSED`);
                    passed++;
                } else {
                    console.log(`  ❌ ${entity} UPDATE: FAILED`);
                    failed++;
                }

                // Test Delete
                console.log(`  🔄 Testing ${entity} DELETE...`);
                const deleteResult = await UnifiedCRUDService.delete(entity, createResult.id);

                if (deleteResult.success) {
                    console.log(`  ✅ ${entity} DELETE: PASSED`);
                    passed++;
                } else {
                    console.log(`  ❌ ${entity} DELETE: FAILED`);
                    failed++;
                }

            } else {
                console.log(`  ❌ ${entity} CREATE: FAILED`);
                failed++;
            }

        } catch (error) {
            console.log(`  ❌ ${entity}: ERROR - ${error.message}`);
            failed++;
        }

        console.log('');
    }

    // Summary
    console.log('📊 QA Test Summary:');
    console.log(`  ✅ Passed: ${passed}`);
    console.log(`  ❌ Failed: ${failed}`);
    console.log(`  📦 Total Requests Captured: ${capturedRequests.length}`);

    console.log('\n🔍 Network Evidence:');
    capturedRequests.forEach((req, i) => {
        console.log(`  ${i+1}. ${req.method} ${req.url}`);
        if (req.body) {
            console.log(`     Payload: ${JSON.stringify(req.body).substring(0, 100)}...`);
        }
    });

    console.log(`\n📝 Logger evidence saved to: ${logPath}`);

    return failed === 0;
}

// Run the test
runStage2QA().then(success => {
    console.log(`\n${success ? '🎉' : '💥'} Stage 2 QA Test ${success ? 'PASSED' : 'FAILED'}`);
    process.exit(success ? 0 : 1);
}).catch(error => {
    console.error('💥 Test execution failed:', error);
    process.exit(1);
});
